### ISSUE TITLE
[Feature request]: Support applications with internal sandboxes

### ISSUE THREAD

**Issue by cjao** [flatpak/flatpak#5921]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

The Flatpak sandbox can create an effective boundary between the running application and the rest of the system. However, it creates problems for Chromium, Firefox, and derived applications that use sophisticated internal sandboxes to isolate processes from each other. These internal partitions are critical as they prevent browser tabs from potentially snooping on each other.

On Linux, Chromium's sandbox was initially implemented using a SetUID helper but nowadays prefers unprivileged [user namespaces and PID namespaces](https://chromium.googlesource.com/chromium/src/+/0e94f26e8/docs/linux_sandboxing.md). However, it is well-known that Chromium does not work out of the box with Flatpak since Flatpak disallows applications from creating user namespaces. Flatpaks of Chromium and its derivatives work around this by [patching the Chromium sandbox](https://github.com/flathub/org.chromium.Chromium/issues/337). This modification not only has not undergone any formal design process or upstream security review but also empirically [weakens](https://github.com/uazo/cromite/issues/1053#issuecomment-2194129916) Chromium's security compared to the [usual RPM/DEB distributions of Chromium](https://discuss.privacyguides.net/t/does-flatpak-weaken-chromium-firefoxs-sandbox/13373/7). Until this is resolved, users of RPM/DEB-based Linux distributions would be ill-served by being advised to adopt the Flatpak versions of Chrome and Firefox because web browsers have the largest attack surface of all user-facing applications.

As a short-term fix, would Flatpak consider adding a `--allow-userns` option to allow give trusted applications sufficient privileges to configure their own sandboxes? While the option to allow user namespaces is potentially a security risk, Flatpak already allows one to poke other holes in the sandbox, such as `--filesystem=host`, because we don't yet live in a world where all applications can run fully confined and use only the portal APIs to interact with the outside world. Ubuntu snap already runs Chromium and Firefox under a [custom security profile](https://snapcraft.io/docs/browser-support-interface) that whitelists user namespace creation to accommodate the browser's internal sandbox. 

While one might ideally want to convince Chromium developers to change their sandbox implementation to accommodate Flatpak, this would likely be a tough sell because Chromium's existing sandbox is already battle-tested and won't stop working anytime soon.

**Comment by Erick555** [comment]
In snap there is one central store where risky permissions are granted by [reviewers committee](https://forum.snapcraft.io/c/store-requests/19). In flatpak anyone can setup a store and grant app any permission they want so there not such thing like giving permission for just _trusted_ apps.

Among dozens of duplicate topics among dozens of repos there is fresh [comment](https://github.com/flatpak/flatpak/issues/5879#issuecomment-2255568180) about why flatpak won't do something to compromise it's basic threat model.

The way to go is to look at how chromium patching can be improved in a way it's acceptable upstream but it's something no one show real interest in to date.

Nothing really changed over years in this topic.

**Comment by cjao** [comment]
The [comment](https://github.com/flatpak/flatpak/issues/5879#issuecomment-2255568180) explains the current state of affairs well but does not contradict the main feature request here which is to provide an on-ramp for Chromium/Firefox to incrementally adopt Flatpak without degrading their own security guarantees. A switch like `--allow-userns` would be in the spirit of the existing sandbox controls like `--filesystem=home` or `--filesystem=host`, which would on one hand allow a malicious app to vacuum up the user's SSH keys and other sensitive data on one's filesystem, but on the other hand make Flatpak a viable bundling mechanism for many existing applications without requiring potentially invasive redesigns.

**Comment by Erick555** [comment]
For me the last two comments clearly contradict with this feature request which is essentially same as the closed issue. Blocking user namespaces is fundamental to flatpak & portal security design unlike protecting ssh keys.

**Comment by Erick555** [comment]
sandbox doesn't protect `~/.ssh` when you pass --filesystem=~/.ssh but it doesn't affect rest of it.

**Comment by Erick555** [comment]
I don't know what you mean - can you explain?

**Comment by Mikenux** [comment]
Having permission to allow unprivileged user namespaces as a short term solution is not a solution because any app will be able to use it. Flatpak apps can be distributed outside of Flathub, so perhaps without proper review (do even Flathub reviewers do a thorough review of apps?). To my knowledge, the current rating system that is shown is only based on how insecure a given permission might be, not on additional evaluations of a repository.

**Comment by rusty-snake** [comment]
xdg-dbus-proxy does not run in a sandbox (I has access to a "execute this shell command please" socket after all). It does however run in a bubblewrap-container.

While x-d-p reads the .flatpak-info from the dbus peer (which is xdg-dbus-proxy) other software still reads from the flatpaked process. This includes software running inside of the flatpak (e.g. glib/gio) and software on the outside. Pipewire does this AFAIK. Generally everything where the flatpak talks to directly (e.g. via `socket=`) can be a consumer of .flatpak-info from the flatpak.

**Comment by rusty-snake** [comment]
So to summarize a few thoughts:
- disabling user-ns by default is good because of increased attack surface to the kernel
- allowing user-ns for some program can be of use
- the problem here is that we must do it in a way that does not violate flatpak's security model
  - this means an application must not be able to modify its vfs to change/hide/... `/.flatpak-info`.
  - there's nothing else I'm aware of that is gated by user-ns
- currently this is blocked by
  1. disabling user-ns (and hence capabilities) via `--disable-userns`
  2. disabling user-ns (and hence capabilities) via seccomp-bpf
  3. disable all mount related syscalls via seccomp-bpf
- While #4187 has the goal to drop all seccomp filtering for performance, this issue is about allowing user-ns functionality. So if we can somehow prove that allowing unshare/clone calls with `CLONE_NEWUSER` without opening a way to modify `/.flatpak-info` we have a plan.

What options to we have?

- Restrict the process with Landlock, this will deny all later attempts to execute mount-syscalls.
  - Good: Does also restrict potential new mount-syscall in the future
  - Bad: It is listed under "Current Limitation" and might be allow in the future
  - Bad: Does not support older kernels
  - Bad: Is less flexible
  - Bad: Does not restrict `chroot`
- Deny all mount-syscalls including `chroot` via seccomp-bpf
  - Good: Does handle all current mount-syscalls including chroot
  - Bad: Does not handle future syscalls
  - Bad: Is seccomp-bpf
- Make `CLONE_NEWUSER` not grant `CAP_SYS_ADMIN` and `CAP_SYS_CHROOT`.
  - Bad: How to do this (without SELinux/AppArmor LSM)?
- Migrate away from `/.flatpak-info` for process identification.
  - Probably the best solution for the future and adoption of other sandboxes by x-d-p.
  - How can this be done? How can a process and all it's childred securely marked as `foo`? CGroups?

**Comment by baaannnaaannn** [comment]
VERY important thing for flatpaks I think
Bcs standard devs use not only for secure focused things but games and other stuff too

**Comment by mcatanzaro** [comment]
So on the one hand, @rusty-snake that is some seriously good brainstorming. I had not considered any of those options.

What I am missing here is justification for why the web browser absolutely needs to be able to create a new user namespace for its custom sandbox in the first place. Of course this permission is obviously required when running outside flatpak, but with flatpak you can simply get a new sandbox for free via flatpak-spawn. The new sandbox is entirely independent of the original sandbox and should accomplish the goal of isolating processes from each other. Why _specifically_ are the Firefox and Chrome developers unable to achieve their goals using flatpak-spawn? Unless there is a very strong reason, I think we should just close this request and stop worrying about it.

Notably, WebKit has succeeded in using flatpak-spawn (at least, nobody has reported a sandbox escape yet) for sandboxing, and WebKit's security model is probably identical to Firefox and Chrome: web processes are assumed to be compromised by malicious user input and actively attempting to attack each other by any means possible.

Reading the [linked issue report](https://github.com/uazo/cromite/issues/1053) I see comments about how Flatpak improves the browser process sandbox but weakens the web content process sandbox. I only skimmed that issue report and I don't see any technical explanation for why, but I agree because the Flatpak sandbox allows more syscalls than the Chromium sandbox and does not attempt to limit kernel attack surface, unlike the Chromium sandbox. This is by design; Flatpak's sandbox is more general purpose and errs towards not breaking legitimate uses of syscalls, and that cannot be changed. But I don't think Flatpak's sandbox prevents installation of new seccomp filters to make things stricter, does it? Chromium could easily impose the same syscall limitations that it uses outside Flatpak if it wished to do so, right, making things stricter than Flatpak's usual sandbox? Am I wrong?

P.S. Certainly an `--allow-userns` is not any worse than `--filesystem=`, but I'd still much rather not add yet another way for apps to ruin their sandboxes without clear benefit.

**Comment by Erick555** [comment]
Firefox cited performance issues if they switch to flatpak-spawn, they need to tweak something first. Chromium devs aren't interested in flatpak support but some time ago they stated 3rd party contribution will be reviewed. Nobody tried to submit anything though. Seccomp filtering works as is for everyone so there are no changes needed for that.

**Comment by rusty-snake** [comment]
> Firefox cited performance issues if they switch to flatpak-spawn, they need to tweak something first.

To expand on this, currently Firefox just forks (`clone(CLONE_...)`) which gives them CoW for all the memory pages. If they would now go and replace the fork with `flatpak-spawn`, the memory usage would explode (with many open tabs). Their plan is to replace this with an ipc call to their new forkserver. Bugzilla ticket: https://bugzilla.mozilla.org/show_bug.cgi?id=1756236

The current state is that the forkserver got recently enabled in Firefox Nigthly. If this does not create major problems / problems outside of the control of Mozilla (i.e. third-party program that "hijack" Firefox or drivers), I guess it will roll down to Firefox Stable by mid to end of this year. After that the plan is to integrate flatpak-spawn in this new IPC aware architecture.

Overall I guess it is more likely IMHO that Firefox finishes the transition to flatpak-spawn that flatpak implementing `--allow-userns` in a stable release that is used by Debian Stable (2028?).

> Chromium 

I never looked deeply into.

> why the web browser

or others program. Are there other programs that want this too?

**Comment by takluyver** [comment]
> it also sounds like there is a perfectly acceptable working solution for Chromium, which just needs to be upstreamed.

I think this might be optimistic. There's a [sizeable patch](https://github.com/flathub/org.chromium.Chromium/blob/master/patches/chromium/flatpak-Add-initial-sandbox-support.patch) that makes the open source Chromium *work* in Flatpak, but there are [questions](https://github.com/flathub/org.chromium.Chromium/issues/337) over whether this weakens Chromium's own sandboxing. I don't know if it does, but that concern is likely to be an issue unless someone can provide a detailed argument that it doesn't.

Meanwhile, Chromium developers [do not seem very interested](https://issues.chromium.org/issues/40928753). In August 2023, one of them said:

> We don't officially support flatpak or snap ... we likely will not officially support a second sandboxing system for Linux.

The real kicker here is that a lot of other apps are now built on Chromium, via Electron or CEF, including proprietary apps which we can't patch, and they inherit this incompatibility even if they don't really need Chromium's sandboxing. The workaround for most is [Zypak](https://github.com/refi64/zypak), which is an impressive piece of work, but also a brittle hack. It often doesn't work very well - e.g. large parts of the Zoom flatpak using CEF don't work - and it's difficult to debug problems when they occur. It's totally possible that some change in Chromium could make a whole load of apps non-viable as Flatpaks over the space of a few weeks/months.

I don't know what the solution is here, but I really think there needs to be some kind of effort to find one, not just "move on".

**Comment by mcatanzaro** [comment]
The Flatpak sandbox is definitely weaker than Chromium's since its seccomp filters are less restrictive. But we've already established this doesn't restrict Chromium from applying its own stricter restrictions, so that shouldn't be a problem.

I'm looking for technical justifications for why user namespace creation is required, not political/social justifications. "Chromium developers don't care about Flatpak" and "many apps depend on Chromium" is not persuasive. I don't think a technical justification actually exists, which is why I suggest closing this issue.

Even if there is some technical reason why it's better for Chromium to create its own sandbox rather than using flatpak-spawn, it's still probably better to use flatpak-spawn anyway so that software centers and app stores can programatically verify that the app is sandboxed. Any application using `--enable-userns` is surely going to be labeled Probably Unsafe; this is not a desirable path forward for applications using Chromium.

**Comment by rusty-snake** [comment]
Moving portals to a better approach than `/.flatpak-info` is much appreciated, good and great work. Nevermind it does not help us here. First, portals do not use `/.flatpak-info` of the flatpak but of xdg-dbus-proxy. Second, other users of `/.flatpak-info` need to migrate to something else as well.

For such other users that do not use D-Bus but an own UNIX-socket, we either still need guaranteed `/.flatpak-info` or a way to reliable/unescapable detect a process as flatpak. Ideas for this are
- SELinux/AppArmor LSM: Does more focus on system-wide admin policies than dynamic unprivileged policies
- Hierarchical namespaces (user and pid) and the `NS_GET_PARENT` ioctl
- CGroup membership
- The ancestor processes

**Comment by cjao** [comment]
>I'm looking for technical justifications for why user namespace creation is required, not political/social justifications. "Chromium developers don't care about Flatpak" and "many apps depend on Chromium" is not persuasive. I don't think a technical justification actually exists, which is why I suggest closing this issue.

What about "Don't fix what ain't broken?" Since Chromium's sandbox has withstood the test of time, any proposal for a redesign bears the burden of justification. New code is [generally more likely than old code to have vulnerabilities](https://security.googleblog.com/2024/09/eliminating-memory-safety-vulnerabilities-Android.html).

**Comment by mcatanzaro** [comment]
> anyway: from the pages that @rusty-snake linked, we see @swick is working on a patch that will make userns safe by making portals not read the .flatpak-info file. thank you very much @swick!

Where is this? Portals have to read .flatpak-info or WebKit will break, but we can probably change WebKit to use any newer mechanism.

**Comment by cjao** [comment]
P.S. @mcatanzaro Can you create PID namespaces without root privileges in the current namespace?

**Comment by cjao** [comment]
P.S. @mcatanzaro Can you create PID namespaces without root privileges in the current user namespace?

**Comment by rusty-snake** [comment]
> Where is this? Portals have to read .flatpak-info or WebKit will break, but we can probably change WebKit to use any newer mechanism.

https://github.com/flatpak/flatpak/pull/5656

**Comment by rusty-snake** [comment]
> Can you create PID namespaces without root privileges in the current user namespace?

root privileges are never required to unshare a namespace. Instead you need `CAP_SYS_ADMIN` in your current user namespace (or you unshare the user namespace at the same time).

This the same capability that gates mounting and one thousand other things. In the initial user-namespace it is almost identical to root (uid 0).

If it would be just `CAP_CLONE_NAMESPACE` ...

**Comment by cjao** [comment]
>I'm looking for technical justifications for why user namespace creation is required, not political/social justifications. "Chromium developers don't care about Flatpak" and "many apps depend on Chromium" is not persuasive. I don't think a technical justification actually exists, which is why I suggest closing this issue.

The so-called "layer-1" sandbox uses a combination of network, mount, and PID namespaces. A SetUID helper was previously used to [set up those namespaces](https://chromium.googlesource.com/chromium/src/+/0e94f26e8/docs/linux_suid_sandbox.md). User namespaces allows Chromium to configure the other types of namespaces without true root or `CAP_SYS_ADMIN`.

**Comment by rusty-snake** [comment]
> User namespaces allows Chromium to configure the other types of namespaces without [] `CAP_SYS_ADMIN`.

User namespaces are exactly the mechanism Chromium uses to gain `CAP_SYS_ADMIN`.

Technically it would be possible to do something like `flatpak-spawn --cap-sys-admin` which then passed `--cap-add CAP_SYS_ADMIN` to bwrap. Hence chromium re-using bubblewrap as possible SetUID helper.

Nevermind this 1. has to be done w/o breaking security guarantees and 2. is not worth the effort if chromium developers are not interested. As long as chromium developers do not give green light upstream for "we consider the technical needs of flatpak for changes in the future" even a contributed implementation upstream or a downstream maintained patch can fail at any time.

> sandbox uses a combination of [] mount [] namespaces. 

Allowing a flatpak to create mount namespaces in the next topic.

**Comment by cjao** [comment]
Another consideration is that by far the largest deployment of the Chromium sandbox on Linux is Chrome OS, not desktop Linux, and even on desktop Linux the vanilla DEB and RPM packages work perfectly well. From a product management point of view it would be hard to justify redesigning the sandbox just to cater to a new and quirky distribution mechanism on desktop Linux.

**Comment by Erick555** [comment]
It's not about redesigning chromium sandbox but adding some _if flatpak then call flatpak-spawn blah blah_. Outside flatpak  chromium sandbox would be intact. Right now chromium under flatpak either fails to start or can be run with `--no-sandbox` which isn't great choice.

**Comment by swick** [comment]
Even though I have a plan to user namespaces available inside flatpak, this might take some time to implement, and much more time to propagate, so browsers *should* adapt to flatpak-spawn sub-sandboxing.

**Comment by mcatanzaro** [comment]
> The so-called "layer-1" sandbox uses a combination of network, mount, and PID namespaces. A SetUID helper was previously used to [set up those namespaces](https://chromium.googlesource.com/chromium/src/+/0e94f26e8/docs/linux_suid_sandbox.md). User namespaces allows Chromium to configure the other types of namespaces without true root or CAP_SYS_ADMIN.

The Flatpak sandbox already uses all of the above namespaces, so I don't see any valid technical reason for Chromium to do so on its own.

**Comment by Erick555** [comment]
The valid technical reason is they do it per process i.e. processes which don't need network are in isolated network namespace without network access, processes which don't need fs access have isolated mount namespace with empty filesystem and so on. Flatpak only does one size fits all.

**Comment by rusty-snake** [comment]
@Erick555 that's missing why *they* need to create these namespaces rather thab flatpak-spawn. If flatpak-spawn misses a feature (e.g. complete empty filesystem) it would be better to improve this API.

**Comment by rusty-snake** [comment]
(A statically linked executable without any dependencies, including non-ELF dependencies like configuration files, could be executed with `execveat`).

**Comment by Erick555** [comment]
> [@Erick555](https://github.com/Erick555) that's missing why _they_ need to create these namespaces rather than flatpak-spawn. If flatpak-spawn misses a feature (e.g. complete empty filesystem) it would be better to improve this API.

It was explained before why they don't use flatpak-spawn and it wasn't part of the question I answered:

> Chromium devs aren't interested in flatpak support

If we're going to do second round of same question-answer then maybe it's time to lock the thread.

**Comment by Erick555** [comment]
Yes flatpak-spawn exist and flathub chromium is actively using it for ages through downstream patch. But some people don't trust downstream patches, they want it upstream. But upstreaming may require a lot of effort. Nobody offered to make that effort after initial skepticism from chromium devs. Then some people demand to tear down flatpak sandbox instead. End of story.

**Comment by Mikenux** [comment]
Shouldn't the creation of a new process with a different sandbox be independent of the sandbox system used? As flarpak-spawn is specific to Flatpak...

**Comment by Erick555** [comment]
The independent way to create sandbox are user namespaces but flatpak blocks them.

**Comment by sbrivio-rh** [comment]
> It's indeed not possible to get a completely empty filesystem using flatpak-spawn. I can't think of any good reason a strong sandbox would require an empty filesystem, though.

We want [that](https://passt.top/passt/tree/isolation.c?id=7c33b1208632a9581d0ee7aabd1e0584a5d1fb20#n293) in passt(1), where we don't need (hence want) access to any filesystem after we set up things.

We just open a couple of files (log file if configured, UNIX domain socket for the guest, standard streams if running in foreground) and then remount `/` from an empty tmpfs with a block limit of 0, and 2 inodes maximum (for ppc64 compatibility).

Note that passt(1) and pasta(1) provide user-mode connectivity to (assumed untrusted) guests and containers, so you really don't want them to have access to any filesystem if not needed. Not even an empty one you can write something to.

For that reason they can't be [packaged](https://github.com/flathub/org.virt_manager.virt_manager.Extension.Qemu/pull/17) as Flatpak, so if somebody sticks to Flatpak, they will be forced to use libslirp/slirp4netns instead, which are surely not as focused on security (not by design, at least).

If Flatpak provides equivalent functionality (e.g. via `flatpak-spawn` commands, or some kind of API, but it needs to be quite [fine-grained](https://passt.top/passt/tree/isolation.c?id=7c33b1208632a9581d0ee7aabd1e0584a5d1fb20#n16)), I'll be more than happy to implement and maintain the whole adaptation upstream.

What I'm surely not happy about and won't support is to disable sandboxing "because it's good enough anyway". Because 1. it's not quite the same thing and 2. because that opens up a vector where somebody can trick them into thinking that they are sandboxed while they aren't.

**Comment by sbrivio-rh** [comment]
> Make `CLONE_NEWUSER` not grant `CAP_SYS_ADMIN` and `CAP_SYS_CHROOT`.
> 
> * Bad: How to do this (without SELinux/AppArmor LSM)?

By the way, I'm working (more off than on, but I should have some time allocated this year) on a project, based on seccomp-bpf and user notifications, that would be a nice fit for these needs, and would probably cover this whole "feature request" in a versatile, declarative way. You could do all that in userspace, which is _a bit_ more flexible.

It's seitan (https://seitan.rocks), the [slides](https://seitan.rocks/static/seitan_devconf_2023.pdf) should provide a quick overview.

I guess it won't be ready in a time frame that's compatible with this pressing (I guess?) use case, though.

**Comment by Mikenux** [comment]
@Erick555: I wanted to mean to use something like a portal because I see that flatpak-spawn is specific to flatpak with its flags. But maybe it's not important?

**Comment by Erick555** [comment]
If something works everywhere but flatpak then you don't need to reinvent general solution, just fix flatpak case.

**Comment by orowith2os** [comment]
I saw landlock mentioned here and would like to mention that I intend on making use of both it and flatpak-spawn in Bottles to improve our in-Flatpak (sub-)sandbox for individual Bottles.

If we wanted to support other environments such as Nix and Snap with our sandbox, they can provide a helper utility that exposes a Flatpak-compatible subsandbox API.

**Comment by BernardoGomesNegri** [comment]
Was taking a look at the proposed org.freedesktop.DBus.Containers API (https://gitlab.freedesktop.org/dbus/dbus/-/commit/39262d0a2913fc8ee951beb3d0241720abf651c0), and was thinking it might help the problem.

For services to identify contained apps, they would expose a DBus API, and when that API is called, the service checks to see what DBus container the DBus sender originates from. 

So Pipewire could expose a DBus API that simply checks the container status of the DBus sender and returns a sandboxed socket.

The only problem is that it would necessitate those other services to use DBus if they wanted to support exposing services in a containerized manner.

**Comment by BernardoGomesNegri** [comment]
Either that or directly parse the cgroup path in /proc/$PID/cgroup of the process, which would arguably be simpler, but flatpak would need to know how to put each application in its own cgroup, and establish a consistent naming format that includes the flatpak instance ID.

Right now it looks like Flatpak apps consistently get a cgroup named "app-flatpak-$APPID-$INSTANCEID.scope" on my system but I do not know if this is an official API or not.

**Comment by smcv** [comment]
> Either that or directly parse the cgroup path in /proc/$PID/cgroup of the process, which would arguably be simpler, but flatpak would need to know how to put each application in its own cgroup, and establish a consistent naming format that includes the flatpak instance ID.


The official way to find out a Flatpak app's instance ID is to parse its `/proc/$PID/root/.flatpak-info`.

> Right now it looks like Flatpak apps consistently get a cgroup named "app-flatpak-$APPID-$INSTANCEID.scope" on my system but I do not know if this is an official API or not.

This is currently "best-effort" - `flatpak run` will try to do this, but if it can't (either because you aren't using `systemd --user`, or because Flatpak can't contact it for whatever reason, or if there is an error) then Flatpak will just continue.

The `$INSTANCEID` part is also not guaranteed: it's just any unique thing (as an implementation detail, in older Flatpak versions it's the process ID and in newer versions it's the instance ID).

If the cgroup was going to be load-bearing and security-critical for something, then we would have to make `flatpak run` fail if it could not do this setup, which would make `systemd --user` a required dependency. Until now, Flatpak has not required systemd.

**Comment by swick** [comment]
FWIW, manual cgroup management is also possible when systemd is not available.

**Comment by radiumatic** [comment]
Has anyone looked into the Android Sandbox API? Chromium does not have access to user-ns on Android either. It uses this instead. It also has Zygote support. (i.e it forks from the Zygote process then sets up the sandbox) and so shouldn't have the performance issues mentioned. 
If the sandbox API is shared between Snaps and Flatpak and the flatpak implementation is more hardened, the Chromium team shouldn't really have any objections?

**Comment by travankor** [comment]
firefox still doesn't use flatpak spawn
